libraries/abstractions/pkcs11/mbedtls/iot_pkcs11_mbedtls.c

```diff
@@ -549,307 +895,1078 @@ CK_DEFINE_FUNCTION( CK_RV, C_OpenSession )( CK_SLOT_ID xSlotID,
-     */
-    if( CKR_OK == xResult )
-    {
-        pxSessionObj->xOperationInProgress = pkcs11NO_OPERATION;
-    }
+                break;
 
-    if( ( NULL != pxSessionObj ) && ( CKR_OK != xResult ) )
-    {
-        vPortFree( pxSessionObj );
+            case ( CKA_EC_POINT ):
+                /* The first 2 bytes are for ASN1 type/length encoding. */
+                lMbedReturn = mbedtls_ecp_point_read_binary( &pxKeyPair->grp, &pxKeyPair->Q, ( ( uint8_t * ) ( xAttribute.pValue ) + 2 ), ( xAttribute.ulValueLen - 2 ) );
+
+                if( lMbedReturn != 0 )
+                {
+                    xResult = CKR_ATTRIBUTE_VALUE_INVALID;
+                }
+
+                break;
+
+            default:
+                PKCS11_PRINT( ( "Unsupported attribute found for EC public key. %d \r\n", xAttribute.type ) );
+                xResult = CKR_ATTRIBUTE_TYPE_INVALID;
+                break;
+        }
     }
 
     return xResult;
 }
 
-/**
- * @brief Terminate a session and release resources.
- */
-CK_DEFINE_FUNCTION( CK_RV, C_CloseSession )( CK_SESSION_HANDLE xSession )
-{   /*lint !e9072 It's OK to have different parameter name. */
+/* Helper function for importing public keys using
+ * C_CreateObject. */
+CK_RV prvCreatePublicKey( CK_ATTRIBUTE_PTR pxTemplate,
+                          CK_ULONG ulCount,
+                          CK_OBJECT_HANDLE_PTR pxObject )
+{
+#define MAX_PUBLIC_KEY_SIZE    3000
+    mbedtls_pk_context xMbedContext;
+    int lDerKeyLength;
+    CK_BYTE_PTR pxDerKey = NULL;
+    CK_KEY_TYPE xKeyType;
     CK_RV xResult = CKR_OK;
-    P11SessionPtr_t pxSession = prvSessionPointerFromHandle( xSession );
+    CK_ATTRIBUTE_PTR pxLabel = NULL;
+    CK_OBJECT_HANDLE xPalHandle = CK_INVALID_HANDLE;
+    CK_BBOOL xPrivateKeyFound = CK_FALSE;
+    mbedtls_pk_init( &xMbedContext );
+    mbedtls_ecp_keypair xKeyPair;
+
+
+    xKeyType = prvGetKeyType( pxTemplate, ulCount );
 
-    if( NULL != pxSession )
+    if( xKeyType == CKK_RSA )
     {
-        /*
-         * Tear down the session.
-         */
+    }
+    else if( xKeyType == CKK_EC ) /* CKK_EC = CKK_ECDSA. */
+    {
+        prvGetLabel( &pxLabel, pxTemplate, ulCount );
 
-        if( NULL != pxSession->xSignKey.pk_ctx )
+        if( CKR_OK != prvGetExistingKeyComponent( &xMbedContext, pxLabel ) )
         {
-            mbedtls_pk_free( &pxSession->xSignKey );
+            mbedtls_ecp_keypair_init( &xKeyPair );
+            xMbedContext.pk_ctx = &xKeyPair;
+            xMbedContext.pk_info = &mbedtls_eckey_info; /* TODO: deprecated ecdsa vs eckey?*/
         }
-
-        if( NULL != pxSession->xSignMutex )
+        else
         {
-            vSemaphoreDelete( pxSession->xSignMutex );
+            xPrivateKeyFound = CK_TRUE;
         }
 
-        /* Free the public key context if it exists. */
-        if( NULL != pxSession->xVerifyKey.pk_ctx )
+        xResult = prvCreateECPublicKey( &xMbedContext, &pxLabel, pxTemplate, ulCount, pxObject );
```

Comment:
Note: There is a memory leak in prvCreateECPublicKey that I am going to work out.   Leaving this comment so I won't forget :)